File: abstract-factory/java/src/VehicleFactory.java
package src;
import src.vehicle.*;
import src.shop.*;
/*
  Vehicle工厂类，继承自抽象工厂基类，生成Vehicle工厂
 */
public class VehicleFactory extends AbstractFactory {
  @Override
  public Vehicle getVehicle(VehicleType type) {
    switch (type) {
      case BUS:
        return new Bus();
      case CAR:
        return new Car();
      case MOTORCYCLE:
        return new Motorcycle();
      case VAN:
        return new Van();
      default:
        return null;
    }

  }

  @Override
  public Shop getShop(String name) {
    return null;
  }
}

File: abstract-factory/go/src/VehicleFactory.go
package src

import (
  "microwind/src/shop"

  "microwind/src/vehicle"
)

const (
  VehicleType_BUS        = "bus"
  VehicleType_CAR        = "car"
  VehicleType_MOTORCYCLE = "motorcycle"
  VehicleType_VAN        = "van"
)

type VehicleFactory struct {
}

func (v *VehicleFactory) GetVehicle(vehicleType string, name string) vehicle.Vehicle {
  switch vehicleType {
  case VehicleType_BUS:
    return &vehicle.Bus{
      Name: name,
    }
  case VehicleType_CAR:
    return &vehicle.Car{
      Name: name,
    }
  case VehicleType_MOTORCYCLE:
    return &vehicle.Motorcycle{
      Name: name,
    }
  case VehicleType_VAN:
    return &vehicle.Van{
      Name: name,
    }
  default:
    return nil
  }
}

func (v *VehicleFactory) GetShop(shopType string, name string) shop.Shop {
  // ignored, this function for Shop
  return nil
}
